src/mds/MDCache.cc

```diff
@@ -7655,12 +7655,10 @@ int MDCache::path_traverse(MDRequestRef& mdr, Message *req, MDSInternalContextBa
   bool null_okay = (onfail == MDS_TRAVERSE_DISCOVERXLOCK);
   bool forward = (onfail == MDS_TRAVERSE_FORWARD);
 
-  assert(mdr || req || fin);
-  assert(!forward || mdr || req);  // forward requires a request
+  assert(mdr || req || fin || !forward);
```

Comment:
These aren't the same, logically. If we have a fin this passes even if forward is set, which is incorrect. You could just stick the logical conditions together with an AND if you want to reduce the calls to assert(), or maybe

```assert( (!forward && fin) || mdr || req );```

if I'm calculating properly. But that'll require a comment explaining itself.